File: remote/src/main/java/com/kelsos/mbrc/ui/dialogs/SettingsDialogFragment.java
package com.kelsos.mbrc.ui.dialogs;

import android.app.Activity;
import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.text.TextUtils;
import android.view.View;
import android.widget.EditText;
import com.afollestad.materialdialogs.MaterialDialog;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.domain.DeviceSettings;

public class SettingsDialogFragment extends DialogFragment {

  public static final String TAG = "settings_dialog";

  public static final int MAX_PORT = 65535;
  public static final int MIN_PORT = 1;

  public static final String ID = "index";
  public static final String PORT = "port";
  public static final String ADDRESS = "address";
  public static final String NAME = "name";
  public static final String HTTP = "http";

  private EditText hostEdit;
  private EditText nameEdit;
  private EditText portEdit;
  private EditText httpEdit;

  private String currentName;
  private String currentAddress;
  private int currentPort;
  private int currentIndex;
  private int currentHttpPort;

  private SettingsDialogListener mListener;

  public static SettingsDialogFragment newInstance(int index) {
    SettingsDialogFragment fragment = new SettingsDialogFragment();
    Bundle args = new Bundle();
    args.putInt(ID, index);
    fragment.setArguments(args);
    return fragment;
  }

  public static SettingsDialogFragment newInstance(DeviceSettings settings) {
    SettingsDialogFragment fragment = new SettingsDialogFragment();
    Bundle args = new Bundle();
    args.putLong(ID, settings.getId());
    args.putString(NAME, settings.getName());
    args.putString(ADDRESS, settings.getAddress());
    args.putInt(PORT, settings.getPort());
    fragment.setArguments(args);
    return fragment;
  }

  @Override public void onAttach(Activity activity) {
    super.onAttach(activity);

    try {
      mListener = (SettingsDialogListener) activity;
    } catch (ClassCastException e) {
      throw new ClassCastException(activity.toString() + " must implement SettingsDialogListener");
    }
  }

  @NonNull @Override public Dialog onCreateDialog(Bundle savedInstanceState) {
    MaterialDialog.Builder builder = new MaterialDialog.Builder(getActivity());
    builder.customView(R.layout.ui_dialog_settings, false);
    builder.title(R.string.settings_dialog_add);
    builder.positiveText(R.string.settings_dialog_add);
    builder.negativeText(android.R.string.cancel);
    builder.onPositive((dialog, which) -> {
      boolean shouldIClose = true;
      String hostname = hostEdit.getText().toString();
      String computerName = nameEdit.getText().toString();

      if (hostname.length() == 0 || computerName.length() == 0) {
        shouldIClose = false;
      }

      String portText = portEdit.getText().toString();
      String httpText = httpEdit.getText().toString();

      int portNum = TextUtils.isEmpty(portText) ? 0 : Integer.parseInt(portText);
      int httpNum = TextUtils.isEmpty(httpText) ? 0 : Integer.parseInt(httpText);

      if (isValid(portNum) && isValid(httpNum) && shouldIClose) {

        DeviceSettings settings = new DeviceSettings();
        settings.setAddress(hostname);
        settings.setName(computerName);
        settings.setPort(portNum);
        settings.setHttp(httpNum);

        mListener.onDialogPositiveClick(SettingsDialogFragment.this, settings);
        dialog.dismiss();
      }
    });

    final MaterialDialog materialDialog = builder.build();
    final View view = materialDialog.getCustomView();
    hostEdit = (EditText) view.findViewById(R.id.settings_dialog_host);
    nameEdit = (EditText) view.findViewById(R.id.settings_dialog_name);
    portEdit = (EditText) view.findViewById(R.id.settings_dialog_port);

    httpEdit = (EditText) view.findViewById(R.id.settings_dialog_http);
    return materialDialog;
  }

  @Override public void onStart() {
    super.onStart();
    nameEdit.setText(currentName);
    hostEdit.setText(currentAddress);

    if (currentHttpPort > 0) {
      httpEdit.setText(String.format("%d", currentHttpPort));
    }
    if (currentPort > 0) {
      portEdit.setText(String.format("%d", currentPort));
    }
  }

  private boolean isValid(int port) {
    if (port < MIN_PORT || port > MAX_PORT) {
      final MaterialDialog.Builder alert = new MaterialDialog.Builder(getActivity());
      alert.title(R.string.alert_invalid_range);
      alert.content(R.string.alert_invalid_port_number);
      alert.positiveText(android.R.string.ok);
      alert.show();
      return false;
    } else {
      return true;
    }
  }

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Bundle args = getArguments();
    if (args != null) {
      currentIndex = args.getInt(ID);
      currentPort = args.getInt(PORT);
      currentAddress = args.getString(ADDRESS);
      currentName = args.getString(NAME);
      currentHttpPort = args.getInt(HTTP);
    }
  }

  public interface SettingsDialogListener {
    void onDialogPositiveClick(SettingsDialogFragment dialog, DeviceSettings settings);
  }
}

File: app/src/main/java/com/kelsos/mbrc/features/settings/SettingsDialogFragment.kt
package com.kelsos.mbrc.features.settings

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.text.TextUtils
import android.widget.EditText
import androidx.fragment.app.DialogFragment
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.kelsos.mbrc.R

class SettingsDialogFragment : DialogFragment() {
  lateinit var hostEdit: EditText
  lateinit var nameEdit: EditText
  lateinit var portEdit: EditText

  private var saveListener: SettingsSaveListener? = null
  private lateinit var settings: ConnectionSettings
  private var edit: Boolean = false

  private fun setConnectionSettings(settings: ConnectionSettings) {
    this.settings = settings
  }

  override fun onAttach(context: Context) {
    super.onAttach(context)
    require(context is SettingsSaveListener) {
      "Activity must implement fragment's listener."
    }
    saveListener = context
  }

  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
    val dialog =
      MaterialAlertDialogBuilder(requireActivity())
        .setView(R.layout.dialog_settings)
        .setTitle(if (edit) R.string.common_edit else R.string.common_add)
        .setPositiveButton(if (edit) R.string.settings_dialog_save else R.string.common_add) {
            dialog,
            _
          ->
          var shouldIClose = true
          val hostname = hostEdit.text.toString()
          val computerName = nameEdit.text.toString()

          if (hostname.isEmpty() || computerName.isEmpty()) {
            shouldIClose = false
          }

          val portText = portEdit.text.toString()

          val portNum = if (TextUtils.isEmpty(portText)) 0 else Integer.parseInt(portText)
          if (isValid(portNum) && shouldIClose) {
            saveListener?.onSave(
              settings.copy(
                name = computerName,
                address = hostname,
                port = portNum
              )
            )
            dialog.dismiss()
          }
        }.setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.dismiss() }
        .show()

    hostEdit = requireNotNull(dialog.findViewById(R.id.settings_dialog_host))
    nameEdit = requireNotNull(dialog.findViewById(R.id.settings_dialog_name))
    portEdit = requireNotNull(dialog.findViewById(R.id.settings_dialog_port))
    return dialog
  }

  override fun onStart() {
    super.onStart()
    nameEdit.setText(settings.name)
    hostEdit.setText(settings.address)

    if (settings.port > 0) {
      portEdit.setText(getString(R.string.common_number, settings.port))
    }
  }

  private fun isValid(port: Int): Boolean = if (port < MIN_PORT || port > MAX_PORT) {
    portEdit.error = getString(R.string.alert_invalid_port_number)
    false
  } else {
    true
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    if (!edit) {
      settings = ConnectionSettings.default()
    }
  }

  fun interface SettingsSaveListener {
    fun onSave(settings: ConnectionSettings)
  }

  companion object {
    private const val MAX_PORT = 65535
    private const val MIN_PORT = 1

    fun newInstance(settings: ConnectionSettings): SettingsDialogFragment {
      val fragment = SettingsDialogFragment()
      fragment.setConnectionSettings(settings)
      fragment.edit = true
      return fragment
    }
  }
}
